### PR TITLE
feat: accept JSON-formatted Shadowsocks Config

### PR DESCRIPTION
**Pull request by fortuna**
Replaces https://github.com/Jigsaw-Code/outline-go-tun2socks/pull/108

Blocked on https://github.com/Jigsaw-Code/outline-go-tun2socks/pull/111

### CODE REVIEW

File: outline/tun2socks/tunnel_android.go [hunk 1]
```diff
@@ -17,6 +17,8 @@ package tun2socks
 import (
 	"runtime/debug"
 
+	"github.com/Jigsaw-Code/outline-go-tun2socks/outline"
+	"github.com/Jigsaw-Code/outline-go-tun2socks/outline/proxy"
```
Comment (jyyi1) [inline outline/tun2socks/tunnel_android.go:21]: ???

File: outline/internal/encoding/utf8/parser.go
Comment (fortuna) [inline outline/internal/encoding/utf8/parser.go:1]: Perhaps out it under `internal/utf8`? I'm not sure we need the intermediary directory.

File: outline/shadowsocks/client.go [hunk 2]
```diff
@@ -29,20 +34,15 @@ import (
...
-	Password   string
-	CipherName string
-	Prefix     []byte
-}
-
+// [Exported] A Shadowsocks client that can be used by Outline-Apps
 type Client = outline.Client
 
-// NewClient provides a gobind-compatible wrapper for [client.NewClient].
+// [Exported] Create a new Shadowsocks client from a non-nil configuration.
```
Comment (fortuna) [inline outline/shadowsocks/client.go:40]: Go has a very specific style for comments that started with the function name. We should keep following that.

File: outline/shadowsocks/config.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+	Port     uint16 `json:"port"`
+	Password string `json:"password"`
+	Method   string `json:"method"`
+	Prefix   string `json:"prefix"`
+}
+
+// [Exported] ParseConfigFromJSON parses a JSON string `in` as a Config object.
+// The JSON string `in` must match the ShadowsocksSessionConfig interface
+// defined in Outline Client.
+func ParseConfigFromJSON(in string) (*Config, error) {
```
Comment (fortuna) [inline outline/shadowsocks/config.go:46]: I don't see the point of having this exported method.
Why not have a `NewClientFromJSON(configJson string)`?

That Config type is not a good abstraction, and we should avoid it.

**Review by fortuna — COMMENTED** [review]
@jyyi1 FYI

File: outline/internal/encoding/utf8/parser.go
Comment (jyyi1) [reply outline/internal/encoding/utf8/parser.go:1]: We can, I just borrowed idea from the standard [`encoding/base64`](https://pkg.go.dev/encoding/base64) and [`unicode/utf8`](https://pkg.go.dev/unicode/utf8)

File: outline/shadowsocks/config.go (hunk 3 above)
Comment (jyyi1) [reply outline/shadowsocks/config.go:46]: I'm trying to separate the concept of `config` and `client`. JSON is only related `config`, client doesn't need to know how the `config` is created.

But if the config interface is not well defined now, we may introduce `NewClientFromJSON` for now. Though I think we need to define it in SDK.

Comment (jyyi1) [reply outline/shadowsocks/config.go:46]: That makes sense, so I guess the code here will be moved to Outline-Apps repo? If this is the case, we can directly create the client (different than the SDK's client) from JSON.

**Comment by jyyi1** [comment]
@fortuna I cannot add you as the reviewer because you created this PR. Feel free to just comment in the PR. Thanks.

**Comment by fortuna** [comment]
Can you create a new PR?

**Pull request closed without merging**